aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+      "Need " #T " of dimension ",                   \
+      DIM,                                           \
+      " and " #T ".size[",                           \
+      DIM_SIZE,                                      \
+      "] == ",                                       \
+      SIZE,                                          \
+      " but got input to be of shape ",              \
+      T.sizes())
+
+/*
+  Some convenience macros
+ */
+
+#define INSERT_BATCH_DIMENSION(A, SIZE)        \
+  if (A.defined()) {                           \
+    auto new_sizes = A.sizes().vec();          \
+    new_sizes.insert(new_sizes.begin(), SIZE); \
+    A.resize_(new_sizes);                      \
+  }
+
+#define DROP_BATCH_DIMENSION(A)    \
+  if (A.defined()) {               \
+    A.resize_(A.sizes().slice(1)); \
+  }
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline bool all_positive(IntArrayRef& arr) {
+  return std::all_of(
+      arr.begin(), arr.end(), [](int64_t item) { return item > 0; });
+}
+
+#define OUTPUTSIZE(INDEX, DIM)                                       \
+  (input.size((INDEX) + input.dim() - (DIM)) + 2 * pad_size[INDEX] - \
+   (dilation_size[INDEX] * (kernel_size[INDEX] - 1) + 1)) /          \
+          stride_size[INDEX] +                                       \
+      1
+
+void conv_dilated_shape_check(
+    int64_t dim,
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const Tensor& grad_output,
+    const Tensor& grad_input,
+    const Tensor& grad_weight,
+    const Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size,
+    const Tensor& columns) {
+  TORCH_CHECK(
+      kernel_size.size() == dim,
+      "kernel sizes length should be ",
+      dim,
+      ", but got ",
+      kernel_size.size());
+  TORCH_CHECK(
+      stride_size.size() == dim,
+      "strides length should be ",
+      dim,
+      ", but got ",
+      stride_size.size());
+  TORCH_CHECK(
+      dilation_size.size() == dim,
+      "dilations length should be ",
+      dim,
+      ", but got ",
+      dilation_size.size());
+  TORCH_CHECK(
+      pad_size.size() == dim,
+      "pads length should be ",
+      dim,
+      ", but got ",
+      pad_size.size());
+
+  TORCH_CHECK(
+      all_positive(kernel_size),
+      "kernel size should be greater than zero, but got ",
+      kernel_size);
+  TORCH_CHECK(
+      all_positive(stride_size),
+      "stride should be greater than zero, but got ",
+      stride_size);
+  TORCH_CHECK(
+      all_positive(dilation_size),
+      "dilation should be greater than zero, but got ",
+      dilation_size);
+
+  if (bias.defined()) {
+    TORCH_CHECK_DIM_SIZE(bias, 1, 0, weight.size(0));
+    TORCH_CHECK(bias.is_contiguous(), "bias ought to be contiguous");
+  }
+
+  if (grad_weight.defined()) {
+    TORCH_CHECK(
+        grad_weight.dim() == dim + 2,
+        "non-empty ",
```

Comment:
It feels like we lost emptiness checking on `input`

```
  THNN_ARGCHECK(!input->is_empty() && (input->dim() == 4 || input->dim() == 5), 2, input,
                "non-empty 4D or 5D (batch mode) tensor expected for input, but got: %s");
```

You don't actually check that it's non-empty, just that the dimensionality is right.

Reply:
Checking of argument shapes and data structure has been revised now.